include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -1324,6 +1324,292 @@ namespace Differentiation
+     *               = -fe_values[msp_fe].gradient(I, q_point);
+     *
+     *             residual_ad[I] -= (dH_I*B) * JxW;
+     *           }
+     *         }
+     *       }
+     *
+     *       // Add contribution from external sources. If these contributions
+     *       // are also solution dependent then they will also be consistently
+     *       // linearized.
+     *       // Loop over faces and accumulate external contributions into the
+     *       // cell total residual.
+     *       // residual_ad[I] += ...
+     *
+     *       // Register the definition of the cell residual
+     *       ad_helper.register_residual_vector(residual_ad);
+     *     }
+     *
+     *     // Compute the residual values and their Jacobian at the
+     *     // evaluation point
+     *     ad_helper.compute_residual(cell_rhs);
+     *     cell_rhs *= -1.0; // RHS = - residual
+     *     ad_helper.compute_linearization(cell_matrix);
+     *   }
+     * @endcode
+     *
+     * @warning ADOL-C does not support the standard threading models used by
+     * deal.II, so this class should @b not be embedded within a multithreaded
+     * function when using ADOL-C number types. It is, however, suitable for use
+     * in both serial and MPI routines.
+     *
+     * @author Jean-Paul Pelteret, 2016, 2017, 2018
+     */
+    template <enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType = double>
+    class ADHelperResidualLinearisation
+      : public ADHelperCellLevelBase<ADNumberTypeCode, ScalarType>
+    {
+    public:
+      /**
+       * Type definition for the floating point number type that is used in,
+       * and results from, all computations.
+       */
+      using scalar_type =
+        typename ADHelperBase<ADNumberTypeCode, ScalarType>::scalar_type;
+
+      /**
+       * Type definition for the auto-differentiation number type that is used
+       * in all computations.
+       */
+      using ad_type =
+        typename ADHelperBase<ADNumberTypeCode, ScalarType>::ad_type;
+
+      /**
+       * @name Constructor / destructor
+       */
+      //@{
+
+      /**
+       * The constructor for the class.
+       *
+       * @param[in] n_independent_variables The number of independent variables
+       * that will be used in the definition of the functions that it is
+       * desired to compute the sensitivities of. In the computation of
+       * $\mathbf{r}(\mathbf{X})$, this will be the number of inputs
+       * $\mathbf{X}$, i.e. the dimension of the domain space.
+       * @param[in] n_dependent_variables The number of scalar functions to be
+       * defined that will have a sensitivity to the given independent
+       * variables. In the computation of $\mathbf{r}(\mathbf{X})$, this will
+       * be the number of outputs $\mathbf{r}$, i.e. the dimension of the
+       * image space.
+       */
+      ADHelperResidualLinearisation(const unsigned int n_independent_variables,
```

Comment:
Would it be worth mentioning in the documentation that in most use cases, and in particular in the code example you showed above, both the number of independent and dependent variables will equal the number of `dofs_per_cell` for the finite element used?